### PR TITLE
[Grappler] improve HoistCWiseUnaryChainsStage to support Reshape

### PR DESCRIPTION
**Pull request by xinan-jiang**
This is a PR from JIZHI, the AI platform in Tencent.

@ezhulenev 
This allows the reshape nodes in the unary op chain to be processed by the optimizer HoistCWiseUnaryChainsStage.

### CODE REVIEW

**Review by ezhulenev — APPROVED** [review]

**Review by rthadur — APPROVED** [review]

File: tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc [hunk 1]
```diff
@@ -1783,6 +1808,133 @@ class HoistCWiseUnaryChainsStage : public ArithmeticOptimizerStage {
...
+    }
+
+    *return_node = unwrapped_node;
+    return Status::OK();
+  }
+
+  Status UnwrapReshapeOutput(const NodeDef* wrapped_node, const int port,
+      PortNodeMap* reshapes, std::set<string>* ctrl_inputs,
+      NodeDef** return_node) const {
+    CHECK(IsReshape(*wrapped_node));
```
Comment (rthadur) [inline tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc:1864]: `New CHECK or LOG(FATAL) statements inserted in code! Prefer using DCHECK, LOG(DFATAL), or DLOG(FATAL) instead`

File: tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc [hunk 2]
```diff
@@ -1783,6 +1808,133 @@ class HoistCWiseUnaryChainsStage : public ArithmeticOptimizerStage {
...
+      TF_RETURN_IF_ERROR(UnwrapReshapeOutput(wrapped_node, port, reshapes,
+                                             ctrl_inputs, return_node));
+    }
+    return Status::OK();
+  }
+
+  Status UnwrapReshapeInput(const NodeDef* wrapped_node, const int port,
+      PortNodeMap* reshapes, std::set<string>* ctrl_inputs,
+      NodeDef** return_node) const {
+    CHECK(IsReshape(*wrapped_node));
```
Comment (rthadur) [inline tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc:1844]: `New CHECK or LOG(FATAL) statements inserted in code! Prefer using DCHECK, LOG(DFATAL), or DLOG(FATAL) instead`

**Review by rthadur — CHANGES_REQUESTED** [review]
@xinan-jiang can you please check review comments.

**Comment by gbaned** [comment]
@xinan-jiang Can you please check @rthadur's comments and keep us posted. Thanks!

**Comment by gbaned** [comment]
@xinan-jiang  Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Pull request closed without merging**